igvc/src/global_mapper/main.cpp

```diff
@@ -11,56 +11,93 @@
+pcl::octree::OctreePointCloudSearch<pcl::PointXYZRGB>::Ptr octree;
 double searchRadius;
+double octree_resolution;
+double probability_thresh;
 
-void icp_transform(pcl::PointCloud<pcl::PointXYZ>::Ptr input)
+//Probabilistic mapping
+//std::vector<pcl::PointCloud<pcl::PointXYZRGB>::Ptr> pointcloud_list;
+std::list<pcl::PointCloud<pcl::PointXYZRGB>::Ptr> pointcloud_list;
+int queue_window_size;
+
+void icp_transform(pcl::PointCloud<pcl::PointXYZRGB>::Ptr input)
 {
-  if (!firstFrame)
+  if (pointcloud_list.size() < queue_window_size)
+  {
+    pointcloud_list.push_back(input);
+  } 
+  if (pointcloud_list.size() == queue_window_size)
   {
-    pcl::IterativeClosestPoint<pcl::PointXYZ, pcl::PointXYZ> icp;
-    icp.setMaxCorrespondenceDistance(maxCorrDist);
-    icp.setMaximumIterations(maxIter);
-    icp.setInputSource(input);
-    icp.setInputTarget(global_map);
-    pcl::PointCloud<pcl::PointXYZ> Final;
-    pcl::PointCloud<pcl::PointXYZ> new_points;
-    icp.align(Final);
-    for (unsigned int i = 0; i < Final.size(); ++i)
+    //Probabilistic calculations here
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr local_pc = pcl::PointCloud<pcl::PointXYZRGB>::Ptr(new pcl::PointCloud<pcl::PointXYZRGB>());
+    pcl::octree::OctreePointCloudSearch<pcl::PointXYZRGB> local_octree(octree_resolution);
+
+    local_octree.setInputCloud(local_pc);
+
+    std::list<pcl::PointCloud<pcl::PointXYZRGB>::Ptr>::iterator it = pointcloud_list.begin();
+    for (unsigned int i = 0; i < pointcloud_list.size(); i++)
     {
-      pcl::PointXYZ searchPoint(Final[i].x, Final[i].y, Final[i].z);
-      std::vector<int> indices;
-      std::vector<float> distances;
-      if (octree.radiusSearch(searchPoint, searchRadius, indices, distances, 1) == 0)
+      //for (auto point : (*pointcloud_list[i]))
+      for (auto point : (**it))
       {
-        octree.addPointToCloud(searchPoint, global_map);
-        new_points.push_back(searchPoint);
+        //G and B represent the probability, they are the same value
+        //Add a point to the local octree if we haven't seen it before
+        if (!local_octree.isVoxelOccupiedAtPoint(point))
+        {
+          point.r = 255;
+          point.g = 0;
+          point.b = 0;
+          local_octree.addPointToCloud(point, local_pc);
+        } else
+        {
+          //Increase probability of the point in the voxel if we see it again
+          int point_idx;
+          float dist;
+          local_octree.approxNearestSearch(point, point_idx, dist);
+          //If the probability is less than the threshold, increase the probability
+          if (local_pc->points[point_idx].g < probability_thresh * 255.0) {
+            local_pc->points[point_idx].g = local_pc->points[point_idx].g + (int)(255.0 / queue_window_size);
+            local_pc->points[point_idx].b = local_pc->points[point_idx].b + (int)(255.0 / queue_window_size);
+          } else if (!firstFrame && local_pc->points[point_idx].g >= probability_thresh * 255.0 
```

Comment:
I am unsure here. Do you want the point only if it exceeds the threshold or if it reaches it. For example if the threshold was 0.6 and the point showed up 3 times then it would not be added, it would have to show up 4 times due to the if else structure. If that is the intention ignore me